### ISSUE TITLE
motor soft limits consistency between python and epics

### ISSUE THREAD

**Issue by mseaberg** [pcdshub/pcdsdevices#597]
There seem to be some inconsistencies in how hutch-python handles soft limits between python and epics. Changing the attributes "low_limit" or "high_limit" does not change any epics pv's, meaning that if the motors are run from epics these limits would have no effect. In contrast, putting or setting low_limit_travel for instance changes the epics soft limit pv, which in my opinion is the desired behavior. I have found that it is easy to end up in an inconsistent state if one changes the "low_limit" or "high_limit" attributes directly and it seems to me that these should be protected perhaps with an underscore.

## Expected Behavior

Soft limits should be consistent between epics and python.

## Current Behavior

It is possible to end up in an inconsistent state by setting python attributes which have no effect on epics pvs.

## Possible Solution

Implement a python wrapper for "low_limit_travel.put(lim)" similar to old hutch-python "set_lowlim" and "set_hilim" methods, that simply put to the soft limit epics pv's.

## Your Environment

xpp3 from xppopr@xpp-control

**Comment by ZLLentz** [comment]
It's set up this way because it introduces performance penalties when an attribute access is directly connected to a PV, since most python programs assume that setattr/getattr are fast operations and interacting with a PV can time out.

I think we definitely add `set_lowlim` and `set_hilim` shortcuts here.

**Comment by ZLLentz** [comment]
If you think the `high_limit` and `low_limit` attributes are traps, we can change their functionality as well.

**Comment by mseaberg** [comment]
I have no problem with the python attributes, just that the implementation for setting soft limits in python needs some work. When I asked about setting soft limits in python, someone suggested that I just set the high_limit and low_limit attributes directly, which I found was causing contradictions with epics soft limits. I never had any issues with the low_limit and high_limit attributes after I only made changes using low_limit_travel, etc. This is why I think these should be more "protected" since I don't think the user should be setting them from a hutch-python session.

**Comment by ZLLentz** [comment]
Understood. If you think we shouldn't see them in hutch python then we can hide them accordingly.

**Comment by silkenelson** [comment]
I think I need a slow-people explanation here: what I think I should see in the motor python object is if you have hit a physical limit switch. Nothing to set here. Then we have the soft limits, that I would have expected to only be defined in the IOC as the motor record already has soft limits. If there is another set of soft limits, I'd like to understand better why we have those in addition to the soft limits in the motor record. How these are settable I have no opinion on.

**Comment by ZLLentz** [comment]
Not all motors have soft limits in the EPICS space (though most of them do) and `setattr` should never put to PVs. Fake motors need limits too if only for testing purposes.

Would it be better to remove the `setattr` functionality entirely for limits? This isn't an `ophyd` built-in so we can change it however we like. We can have functions for setting limits. We can remove non-EPICS limits for motors that have EPICS limits if that's what you want us to do.

**Comment by mseaberg** [comment]
My opinion is that these attributes should not be allowed to be set manually for motors with EPICS limits since this does (from what I think I observed) effectively cause there to be another set of soft limits that is probably local to the python session. It does seem that having a function to set limits is better since it's easier to prevent someone from breaking the desired behavior that way.

**Comment by silkenelson** [comment]
I personally would prefer that the python-motor would use python limits where necessary and EPICS/motor record limits where possible and that be invisible to the user. If we have soft limits in python AND soft limits in EPICS, then we have a disconnect when we e.g. open the motor-screen and see different limits from what we just applied in beam line python. I admittedly do to know how hard/easy that is and if what I'd see as natural reflects what the common user would expect to happen (Matt is our best proxy here).

**Comment by mseaberg** [comment]
Yes, Silke, that is my main concern, that someone would think they are setting a limit in python but that this wouldn't be reflected in the edm screen.

**Comment by cristinasewell** [comment]
This is one of the Issues on my bucket list and I know we tackled some things related to it in #707 - I'm wondering if the users experienced any issues with those changes or if those changes were enough to solve the problem of soft limits consistency between python and epics?
